### PR TITLE
Loader subsystem and commands

### PR DESCRIPTION
**Pull request by OhadLevine**

### CODE REVIEW

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 1]
```diff
@@ -7,21 +7,22 @@
  * The RobotConstants maps constants to a variable name.
  */
 public abstract class RobotConstants extends RobotMap {
-    public DrivetrainConstants drivetrainConstants = new DrivetrainConstants(); 
+    public DrivetrainConstants drivetrainConstants = new DrivetrainConstants();
     public VisionConstants visionConstants = new VisionConstants();
     public ControlConstants controlConstants = new ControlConstants();
-    public MotionProfilingConstants motionProfilingConstants = new MotionProfilingConstants(); 
+    public MotionProfilingConstants motionProfilingConstants = new MotionProfilingConstants();
+    public LoaderConstants loaderConstants = new LoaderConstants();
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/constants/RobotConstants.java]: move loader to after drivetrain and before vision

File: src/main/java/frc/robot/subsystems/loader/Loader.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+   */
+  public Loader() {
+    motor = new WPI_TalonSRX(robotConstants.can.LOADER_TALONSRX);
+    motor.configClosedloopRamp(robotConstants.loaderConstants.RAMP_RATE);
+    motor.configSupplyCurrentLimit(
+        new SupplyCurrentLimitConfiguration(true, robotConstants.loaderConstants.CURRENT_LIMIT,
+            robotConstants.loaderConstants.THRESHOLD_LIMIT, robotConstants.loaderConstants.TIMEOUT));
+  }
+
+  /** @return Meters per second */
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/Loader.java]: change `/** @return rotations per minute */`

File: src/main/java/frc/robot/subsystems/loader/Loader.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  public double getVelocity() {
+    return motor.getSelectedSensorVelocity() * 10 / robotConstants.loaderConstants.TICKS_PER_METER;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/Loader.java]: change to `return motor.getSelectedSensorVelocity() * 600 / robotConstants.loaderConstants.TICKS_PER_ROTATION;`

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * This does PID on the loader based on the desired velocity
+   */
+  public LoaderPID(DoubleSupplier desiredVelocity, BooleanSupplier isFinished, boolean isTuning) {
+    addRequirements(Robot.loader);
+    this.desiredVelocity = desiredVelocity;
+    this.isFinished = isFinished;
+    this.isTuning = isTuning;
+    feedForward = robotConstants.loaderConstants.FEED_FORWARD;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: feedforward is one word. so replace 'feedForward' with 'feedforward'

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 5]
```diff
@@ -41,13 +42,25 @@
...
         public double LIMELIGHT_ANGLE_OFFSET;
         public double TARGET_NOT_FOUND_WAIT_TIME;
     }
 
+    public static class LoaderConstants {
+        public double RAMP_RATE;
+        public double CURRENT_LIMIT;
+        public double THRESHOLD_LIMIT;
+        public double TIMEOUT;
+        public double TICKS_PER_METER;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/constants/RobotConstants.java]: Change to `public double TICKS_PER_ROTATION;`

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 6]
```diff
@@ -41,13 +42,25 @@
...
+        public PIDSettings PID_SETTINGS;
+        public SimpleMotorFeedforward FEED_FORWARD;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/constants/RobotConstants.java]: feedforward is one word. so replace `FEED_FORWARD` with FEEDFORWARD

File: src/main/java/frc/robot/constants/robots/RobotA.java [hunk 7]
```diff
@@ -39,12 +39,23 @@ public RobotA() {
...
-        controlConstants.visionRotationSettings = new PIDSettings(0,0,0,0,0);
+        controlConstants.visionDistanceSettings = new PIDSettings(0, 0, 0, 0, 0);
+        controlConstants.visionRotationSettings = new PIDSettings(0, 0, 0, 0, 0);
+
+        /* Loader Constants */
+        loaderConstants.RAMP_RATE = 0;
+        loaderConstants.CURRENT_LIMIT = 0;
+        loaderConstants.THRESHOLD_LIMIT = 0;
+        loaderConstants.TIMEOUT = 0;
+        loaderConstants.TICKS_PER_METER = 0;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/constants/robots/RobotA.java]: change to ticks per rotation and make it equals 1

File: src/main/java/frc/robot/subsystems/loader/Loader.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.ctre.phoenix.motorcontrol.SupplyCurrentLimitConfiguration;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.subsystems.MoveableSubsystem;
+
+import static frc.robot.Robot.robotConstants;;
+
+public class Loader extends SubsystemBase implements MoveableSubsystem {
+  WPI_TalonSRX motor;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/Loader.java]: change `motor` to talonSRX

File: src/main/java/frc/robot/subsystems/loader/Loader.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+  /** @return Meters per second */
+  public double getVelocity() {
+    return motor.getSelectedSensorVelocity() * 10 / robotConstants.loaderConstants.TICKS_PER_METER;
+  }
+
+  public void setVoltage(double voltage) {
+    motor.setVoltage(voltage);
+  }
+
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/Loader.java:35]: add `@Override`

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  private PIDController pidController;
+  private DoubleSupplier desiredVelocity;
+  private BooleanSupplier isFinished;
+  private boolean isTuning;
+  private SimpleMotorFeedforward feedForward;
+
+  /**
+   * This does PID on the loader based on the desired velocity
+   */
+  public LoaderPID(DoubleSupplier desiredVelocity, BooleanSupplier isFinished, boolean isTuning) {
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: we can just call on a command [withInterrupt](https://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj2/command/Command.html#withInterrupt(java.util.function.BooleanSupplier)). This tells the scheduler to interrupt the command once the boolean condition supplier is true. So we don't really need to have an isFinished supplier

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    this.isFinished = isFinished;
+    this.isTuning = isTuning;
+    feedForward = robotConstants.loaderConstants.FEED_FORWARD;
+  }
+
+  @Override
+  public void initialize() {
+    if (isTuning)
+      SmartDashboard.putData("Loader pid controller: ", pidController);
+    pidController = new PIDController(robotConstants.loaderConstants.PID_SETTINGS.getKP(),
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: create pidController in constructor rather in initialize. and reset it in initialize

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  @Override
+  public void execute() {
+    Robot.loader.setVoltage(
+        MathUtil.clamp(pidController.calculate(Robot.loader.getVelocity(), desiredVelocity.getAsDouble()), -6, 6)
+            + feedForward.calculate(desiredVelocity.getAsDouble()));
+  }
+
+  @Override
+  public void end(boolean interrupted) {
+    pidController.close();
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: remove close

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    Robot.loader.move(0);
+  }
+
+  @Override
+  public boolean isFinished() {
+    return isFinished.getAsBoolean();
+  }
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: remove isFinished

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java (hunk 10 above)
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: Add more constructors that receive a double setpoint and that have isTuning defaulted to 0

File: src/main/java/frc/robot/constants/RobotMap.java [hunk 14]
```diff
@@ -13,29 +13,26 @@
...
   public DIO dio = new DIO();
   public PWM pwm = new PWM();
-  public I2C i2c = new I2C(); 
+  public I2C i2c = new I2C();
 
   // TODO: Set variables for hardware components
-  
   public static class CAN {
-      
+    public int LOADER_TALONSRX;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/constants/RobotMap.java]: change to `public int LOADER_TALON_SRX;`

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    feedForward = robotConstants.loaderConstants.FEED_FORWARD;
+  }
+
+  @Override
+  public void initialize() {
+    if (isTuning)
+      SmartDashboard.putData("Loader pid controller: ", pidController);
+    pidController = new PIDController(robotConstants.loaderConstants.PID_SETTINGS.getKP(),
+        robotConstants.loaderConstants.PID_SETTINGS.getKI(), robotConstants.loaderConstants.PID_SETTINGS.getKD());
+    pidController.setTolerance(robotConstants.loaderConstants.PID_SETTINGS.getTolerance());
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: what the point in tolerance if there isn't any atSetpoint?

**Review by EliyaFishman — CHANGES_REQUESTED** [review]
pls change

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 16]
```diff
@@ -7,21 +7,32 @@
...
     }
-    
+
+    public static class LoaderConstants {
+        public double RAMP_RATE;
+        public double CURRENT_LIMIT;
+        public double THRESHOLD_LIMIT;
+        public double TIMEOUT;
+        public double TICKS_PER_ROTATION;
+        public PIDSettings PID_SETTINGS;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/constants/RobotConstants.java]: PID_SETTINGS not here. in controlConstans

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 17]
```diff
@@ -7,21 +7,32 @@
...
+        public SimpleMotorFeedforward FEEDFORWARD;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/constants/RobotConstants.java]: FEEDFORWARD in controlConstans

File: src/main/java/frc/robot/subsystems/loader/Loader.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+  /**
+   * This class holds all the methods for the Loader which takes the POWER CELLS
+   * from the Mixer and loads it into the Shooter
+   */
+  public Loader() {
+    talonSRX = new WPI_TalonSRX(robotConstants.can.LOADER_TALON_SRX);
+    talonSRX.configClosedloopRamp(robotConstants.loaderConstants.RAMP_RATE);
+    talonSRX.configSupplyCurrentLimit(
+        new SupplyCurrentLimitConfiguration(true, robotConstants.loaderConstants.CURRENT_LIMIT,
+            robotConstants.loaderConstants.THRESHOLD_LIMIT, robotConstants.loaderConstants.TIMEOUT));
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/loader/Loader.java:24]: add coast mode

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+  @Override
+  public void execute() {
+    Robot.loader.setVoltage(
+        MathUtil.clamp(pidController.calculate(Robot.loader.getVelocity(), desiredVelocity.getAsDouble()), -6, 6)
+            + feedforward.calculate(desiredVelocity.getAsDouble()));
+  }
+
+  @Override
+  public void end(boolean interrupted) {
+    Robot.loader.move(0);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: import loader static and use stop move

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+import edu.wpi.first.wpilibj.controller.PIDController;
+import edu.wpi.first.wpilibj.controller.SimpleMotorFeedforward;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import edu.wpi.first.wpiutil.math.MathUtil;
+import frc.robot.Robot;
+
+import static frc.robot.Robot.robotConstants;
+
+public class LoaderPID extends CommandBase {
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: LoaderPID to setLoaderVelocity

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  public LoaderPID(DoubleSupplier desiredVelocity) {
+    this(desiredVelocity, false);
+  }
+
+  @Override
+  public void initialize() {
+    if (isTuning)
+      SmartDashboard.putData("Loader pid controller: ", pidController);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: Loader pid controller: to Loader pid controller

File: src/main/java/frc/robot/subsystems/loader/LoaderPID.java [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+import static frc.robot.Robot.robotConstants;
+
+public class LoaderPID extends CommandBase {
+  private PIDController pidController;
+  private DoubleSupplier desiredVelocity;
+  private boolean isTuning;
+  private SimpleMotorFeedforward feedforward;
+
+  /**
+   * This does PID on the loader based on the desired velocity
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/loader/LoaderPID.java]: * This does PID on the loader based on the desired velocity
in which units vel? and does PID isn't a good explanation

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
change

**Review by HilelAmiel — APPROVED** [review]
done done

**Review by EliyaFishman — APPROVED** [review]

**Pull request merged**